File: problems/Java/115. Distinct Subsequences.java
class Solution {
    public int numDistinct(String s, String t) {
        int m = s.length();
        int n = t.length();
        if (m < n) return 0;
        
        int[] dp = new int[n + 1];
        dp[0] = 1;
        
        char[] sc = s.toCharArray();
        char[] tc = t.toCharArray();
        
        for (int i = 0; i < m; i++) {
            for (int j = n; j >= 1; j--) {
                if (sc[i] == tc[j - 1]) {
                    dp[j] += dp[j - 1];
                }
            }
        }
        return dp[n];
    }
}

File: problems/Python/115. Distinct Subsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        m, n = len(s), len(t)
        if m < n:
            return 0
        
        dp = [0] * (n + 1)
        dp[0] = 1
        
        for sc in s:
            for j in range(n, 0, -1):
                if sc == t[j - 1]:
                    dp[j] += dp[j - 1]
                    
        return dp[n]